src/Fetch.js

```diff
@@ -22,58 +22,53 @@ var Fetch = {
 
 #if FETCH_SUPPORT_INDEXEDDB
   openDatabase: function(dbname, dbversion, onsuccess, onerror) {
-    try {
+    // Run asynchronously so that `onerror` doesn't run synchronously
+    // when the `try` fails (#18329)
+    setTimeout(() => {
+      try {
 #if FETCH_DEBUG
-      dbg('fetch: indexedDB.open(dbname="' + dbname + '", dbversion="' + dbversion + '");');
+        dbg('fetch: indexedDB.open(dbname="' + dbname + '", dbversion="' + dbversion + '");');
 #endif
-      var openRequest = indexedDB.open(dbname, dbversion);
-    } catch (e) { return onerror(e); }
+        var openRequest = indexedDB.open(dbname, dbversion);
+      } catch (e) { return onerror(e); }
 
-    openRequest.onupgradeneeded = (event) => {
+      openRequest.onupgradeneeded = (event) => {
 #if FETCH_DEBUG
-      dbg('fetch: IndexedDB upgrade needed. Clearing database.');
+        dbg('fetch: IndexedDB upgrade needed. Clearing database.');
 #endif
-      var db = /** @type {IDBDatabase} */ (event.target.result);
-      if (db.objectStoreNames.contains('FILES')) {
-        db.deleteObjectStore('FILES');
-      }
-      db.createObjectStore('FILES');
-    };
-    openRequest.onsuccess = (event) => onsuccess(event.target.result);
-    openRequest.onerror = (error) => onerror(error);
+        var db = /** @type {IDBDatabase} */ (event.target.result);
+        if (db.objectStoreNames.contains('FILES')) {
+          db.deleteObjectStore('FILES');
+        }
+        db.createObjectStore('FILES');
+      };
+      openRequest.onsuccess = (event) => onsuccess(event.target.result);
+      openRequest.onerror = (error) => onerror(error);
+    })
   },
 #endif
 
   staticInit: function() {
-    var isMainThread = true;
-
 #if FETCH_SUPPORT_INDEXEDDB
     var onsuccess = (db) => {
 #if FETCH_DEBUG
       dbg('fetch: IndexedDB successfully opened.');
 #endif
       Fetch.dbInstance = db;
-
-      if (isMainThread) {
-        removeRunDependency('library_fetch_init');
-      }
+      removeRunDependency('library_fetch_init');
     };
+
     var onerror = () => {
 #if FETCH_DEBUG
       dbg('fetch: IndexedDB open failed.');
 #endif
       Fetch.dbInstance = false;
-
-      if (isMainThread) {
-        removeRunDependency('library_fetch_init');
-      }
+      removeRunDependency('library_fetch_init');
     };
+
+    addRunDependency('library_fetch_init');
```

Comment:
Given that you have moved this up such that it happens before openDatabase, why do you still need the setTimeout above?

Reply:
The `setTimeout()` is not needed but makes the interface of `openDatabase()` more transparent as all its callbacks are then called asynchronously. I made that change following a suggestion of @kripken. Should I revert the commit?

Reply:
I don't think its needed once we fix the addRunDependency ordering.  So I guess I disageee with @kripken for once.  Why add extra unneeded nesting of code and running of the event loop.

Reply:
Sounds good, now reverted.